src/main/java/org/jenkinsci/plugins/workflow/cps/global/UserDefinedGlobalVariableAction.java

```diff
@@ -0,0 +1,160 @@
+import org.kohsuke.stapler.Stapler;
+import org.kohsuke.stapler.StaplerRequest;
+import org.kohsuke.stapler.export.Exported;
+import org.kohsuke.stapler.export.ExportedBean;
+
+import static hudson.Util.fixNull;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+@ExportedBean(defaultVisibility = 999)
+public class UserDefinedGlobalVariableAction implements RunAction2 {
+
+    private transient Run run;
+    public List<UserDefinedGlobalVariableData> userGlobalVars = new ArrayList<>();
+
+    public UserDefinedGlobalVariableAction() {
+    }
+
+    public UserDefinedGlobalVariableAction(List<UserDefinedGlobalVariableData> globalVars) {
+        for(UserDefinedGlobalVariableData v : globalVars) {
+            userGlobalVars.add(v);
+        }
+    }
+
+    public String getDisplayName() {
+        return "User Defined Global Variables";
+    }
+
+    public String getIconFileName() {
+        return "attribute.png";
+    }
+
+    public String getUrlName() {
+        return "userglobalvars";
+    }
+
+    public void addUserDefinedGlobalVariable(UserDefinedGlobalVariable userVar) {
+        userGlobalVars.add(new UserDefinedGlobalVariableData(userVar));
+    }
+
+    @Exported
+    public List<Map<String,String>> apiUserGlobalVars() {
+        ArrayList<Map<String,String>> apiData = new ArrayList<>();
+        for(UserDefinedGlobalVariableData v : userGlobalVars) {
+            Map<String,String> data = new HashMap<>();
+            data.put("name", v.getName());
+            data.put("library", v.getLibraryName());
+            data.put("version", v.getLibraryVersion());
+            apiData.add(data);
+        }
+        return apiData;
+    }
+
+    public List<UserDefinedGlobalVariableData> getUserGlobalVars() {
+        return userGlobalVars;
+    }
+
+    @Override
+    public String toString() {
+        return super.toString()+"[userGlobalVars="+userGlobalVars+"]";
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+
+        UserDefinedGlobalVariableAction that = (UserDefinedGlobalVariableAction) o;
+
+        return Objects.equals(userGlobalVars, that.userGlobalVars);
+    }
+
+    @Override
+    public int hashCode() {
+        return userGlobalVars.hashCode();
+    }
+
+    /* Package protected for easier testing */
+    static final Logger LOGGER = Logger.getLogger(UserDefinedGlobalVariableAction.class.getName());
+
+    @Override
+    public void onAttached(Run<?, ?> run) {
+        this.run = run;
+    }
+
+    @Override
+    public void onLoad(Run<?, ?> run) {
+        this.run = run;
+    }
+
+    public Run getRun() {
+        return run;
+    }
+
+    public Api getApi() { return new Api(this); }
+
+    public static class UserDefinedGlobalVariableData {
```

Comment:
@j3p0uk I would add `@ExportedBean` to this class and add `@Exported` to its getter methods, and then in the parent class I would delete `apiUserGlobalVars` and add `@Exported` to `getUserGlobalVars` instead.